unittests/finality_proof.hpp

```diff
@@ -0,0 +1,289 @@
+#pragma once
+
+#include "finality_test_cluster.hpp"
+
+using mvo = mutable_variant_object;
+
+namespace finality_proof {
+
+   // data relevant to IBC
+   struct ibc_block_data_t {
+      signed_block_ptr block;
+      qc_data_t qc_data;
+      action_trace onblock_trace;
+      finality_data_t finality_data;
+      uint32_t active_finalizer_policy_generation;
+      uint32_t last_pending_finalizer_policy_generation;
+      uint32_t last_proposed_finalizer_policy_generation;
+      digest_type action_mroot;
+      digest_type base_digest;
+      digest_type active_finalizer_policy_digest;
+      digest_type last_pending_finalizer_policy_digest;
+      digest_type last_proposed_finalizer_policy_digest;
+      digest_type finality_digest;
+      digest_type computed_finality_digest;
+      digest_type afp_base_digest;
+      digest_type finality_leaf;
+      digest_type finality_root;
+   };
+
+   //generate a proof of inclusion for a node at index from a list of leaves
+   static std::vector<digest_type> generate_proof_of_inclusion(const std::vector<digest_type> leaves, const size_t index) {
+      auto _leaves = leaves;
+      auto _index = index;
+
+      std::vector<digest_type> merkle_branches;
+
+      while (_leaves.size()>1){
+         std::vector<digest_type> new_level;
+         for (size_t i = 0 ; i < _leaves.size() ; i+=2){
+            digest_type left = _leaves[i];
+
+            if (i + 1 < _leaves.size() && (i + 1 != _leaves.size() - 1 || _leaves.size() % 2 == 0)){

```

Comment:
It seems `(i + 1 != _leaves.size() - 1 || _leaves.size() % 2 == 0))` is redundant when we know `i + 1 < _leaves.size()`.